### DISCUSSION TITLE
How best to cache series information and history?

### DISCUSSION THREAD

**Discussion by MaxHasADHD** [trakt/trakt-api#372] [Q&A]
I have a TV tracker app that shows when series air, all of their seasons and episodes, and the users history for those series. On sync I fetch (in order):
1. last activities (`sync/last_activities`)
2. hidden shows (`users/hidden/progress_watched`)
3. history (`sync/watched/shows?extended=full`)
4. watchlist (`sync/watchlist/shows?extended=full`)
5. Episode user ratings (`sync/ratings/episodes`)
6. Seasons (`shows/{trakt_id}/seasons?extended=full,episodes`)

This is a lot of data to fetch, especially after most of it has been cached. There isn't much I can do for first sync, and on subsequent syncs I use `last_activities` to figure out if I need to fetch history or the watchlist, etc. Sync can still take a long time though because I fetch the users full history. I do this because the user may mark an episode as watched in the past before the last sync, or marked something as unwatched. I'm also fetching full series seasons, whereas once cached I really only need to refresh the current season and watch out for new seasons. That could improve speed especially for long-running series like game shows, SNL, etc since episodes don't really change once aired.

I would love to know how the Trakt app handles first and subsequent syncs, the best way to fetch history and keep it accurate, and which endpoints to call to optimize fetching episodes and seasons.

**Comment by mawro453** [comment]
I've had a similar issue as you in my discussion: https://github.com/trakt/api-help/discussions/317

To solve the issue, that a user marks an episode as watched in the past, I would suggest adding the ability to query the watched endpoint via the last_update timestamp, with start and end timestamps provided. This way, only the relevant data would be fetched, greatly reducing the amount of data transmitted during the sync process. I believe this improvement would greatly enhance the efficiency and ease of use of the API, and would be of great benefit to many other users as well.

**Reply by rectifyer** [reply]
Not a bad idea, but the issue is plays can be added or removed. That's why we have always relied on returning the full data set so your app can sync it all. We store some limited info about removed plays, but it's not forever so we can't have a method that includes the removals.
